### ISSUE TITLE
"extra fields not permitted" error when using `Extra.forbid` on `dataclass` property that's a `dict` or `list` of another `dataclass`

### ISSUE THREAD

**Issue by DetachHead** [pydantic/pydantic#4343]
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://pydantic-docs.helpmanual.io) and still think this is a bug
- [X] I am confident that the issue is with pydantic  (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

when instanciating a `dataclass` where a property is a `list` or `dict` of another `dataclass`, it causes this error:

> extra fields not permitted (type=value_error.extra)

### Example Code

```Python
from pydantic import Extra
from pydantic.dataclasses import dataclass


class ExtraPropertiesForbidden:
    extra = Extra.forbid


@dataclass(config=ExtraPropertiesForbidden)
class Bar:
    ...


@dataclass(config=ExtraPropertiesForbidden)
class Foo:
    a: list[Bar]


a = Foo(a=[Bar()])  # error: extra fields not permitted (type=value_error.extra)
```

### Python, Pydantic & OS Version
this issue seems to have been introduced after the 1.9.1 release. this occurred after installing the latest commit from master:

```
pip uninstall pydantic
pip install git+https://github.com/samuelcolvin/pydantic.git@a69136d2096e327ed48954a298c775dceb1551d1
```
```Text
pydantic version: 1.9.0
            pydantic compiled: False
                 install path: C:\Users\chungus\AppData\Local\pypoetry\Cache\virtualenvs\project-e01-Hfn1-py3.10\Lib\site-packages\pydantic
               python version: 3.10.6 (tags/v3.10.6:9c7b4bd, Aug  1 2022, 21:53:49) [MSC v.1932 64 bit (AMD64)]
                     platform: Windows-10-10.0.19042-SP0
     optional deps. installed: ['typing-extensions']
```

### Affected Components

- [ ] [Compatibility between releases](https://pydantic-docs.helpmanual.io/changelog/)
- [ ] [Data validation/parsing](https://pydantic-docs.helpmanual.io/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://pydantic-docs.helpmanual.io/usage/exporting_models/) - `.dict()` and `.json()`
- [ ] [JSON Schema](https://pydantic-docs.helpmanual.io/usage/schema/)
- [X] [Dataclasses](https://pydantic-docs.helpmanual.io/usage/dataclasses/)
- [X] [Model Config](https://pydantic-docs.helpmanual.io/usage/model_config/)
- [ ] [Field Types](https://pydantic-docs.helpmanual.io/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://pydantic-docs.helpmanual.io/usage/validation_decorator/)
- [ ] [Generic Models](https://pydantic-docs.helpmanual.io/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://pydantic-docs.helpmanual.io/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [ ] [Settings Management](https://pydantic-docs.helpmanual.io/usage/settings/)
- [ ] [Plugins](https://pydantic-docs.helpmanual.io/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by samuelcolvin** [comment]
Very good catch, thank you @DetachHead.

**Comment by samuelcolvin** [comment]
@PrettyWood would be amazing if you could take a look at this :pray:.

**Comment by PrettyWood** [comment]
Great catch! I'll check tonight

**Comment by DetachHead** [comment]
i've already fixed it in #4354 if you wanna take a look

**Comment by samuelcolvin** [comment]
:see_no_evil: sorry I missed that, will look soon.
